src/storage_stage.rs

```diff
@@ -0,0 +1,450 @@
+        | ((rkey[3] as usize) << 24);
+    res &= NUM_IDENTITIES - 1;
+    res
+}
+
+impl StorageState {
+    pub fn new() -> Self {
+        let storage_keys = Arc::new(RwLock::new(vec![0u8; KEY_SIZE * NUM_IDENTITIES]));
+        let storage_results = Arc::new(RwLock::new(vec![Hash::default(); NUM_IDENTITIES]));
+
+        StorageState {
+            storage_keys,
+            storage_results,
+        }
+    }
+
+    pub fn get_mining_key(&self, key: &Pubkey) -> Vec<u8> {
+        let idx = get_identity_index_from_pubkey(key);
+        self.storage_keys.read().unwrap()[idx..idx + KEY_SIZE].to_vec()
+    }
+
+    pub fn get_mining_result(&self, key: &Pubkey) -> Hash {
+        let idx = get_identity_index_from_pubkey(key);
+        self.storage_results.read().unwrap()[idx]
+    }
+}
+
+impl StorageStage {
+    pub fn new(
+        storage_state: &StorageState,
+        storage_entry_receiver: EntryReceiver,
+        ledger_path: Option<&str>,
+        keypair: Arc<Keypair>,
+        exit: Arc<AtomicBool>,
+        entry_height: u64,
+    ) -> Self {
+        let storage_keys_ = storage_state.storage_keys.clone();
+        let storage_results_ = storage_state.storage_results.clone();
+        let ledger_path = if let Some(ledger_str) = ledger_path {
+            Some(String::from(ledger_str))
+        } else {
+            None
+        };
+        let t_storage_mining_verifier = Builder::new()
+            .name("solana-storage-mining-verify-stage".to_string())
+            .spawn(move || {
+                let exit = exit.clone();
+                let mut poh_height = 0;
+                let mut current_key = 0;
+                let mut entry_height = entry_height;
+                loop {
+                    if let Some(ref ledger_path_str) = ledger_path {
+                        if let Err(e) = Self::process_entries(
+                            &keypair,
+                            &storage_keys_,
+                            &storage_results_,
+                            &storage_entry_receiver,
+                            ledger_path_str,
+                            &mut poh_height,
+                            &mut entry_height,
+                            &mut current_key,
+                        ) {
+                            match e {
+                                Error::RecvTimeoutError(RecvTimeoutError::Disconnected) => break,
+                                Error::RecvTimeoutError(RecvTimeoutError::Timeout) => (),
+                                _ => info!("Error from process_entries: {:?}", e),
+                            }
+                        }
+                    }
+                    if exit.load(Ordering::Relaxed) {
+                        break;
+                    }
+                }
+            }).unwrap();
+
+        StorageStage {
+            t_storage_mining_verifier,
+        }
+    }
+
+    pub fn process_entry_crossing(
+        _storage_results: &Arc<RwLock<StorageResults>>,
```

Comment:
Why include the unused parameters?

Reply:
They are used when `feature = 'cuda'` is set, but unused otherwise.